libkbfs/file_data.go

```diff
@@ -97,6 +107,357 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
+	if !pblock.IsInd {
+		// Return a single empty path and a child map with only this
+		// block in it, under the assumption that the caller already
+		// checked the range for this block.
+		return [][]parentBlockAndChildIndex{nil},
+			map[BlockPointer]*FileBlock{ptr: pblock}, -1, nil
+	}
+
+	type resp struct {
+		pathsFromRoot   [][]parentBlockAndChildIndex
+		blocks          map[BlockPointer]*FileBlock
+		nextBlockOffset int64
+	}
+
+	// Search all of the in-range child blocks, and their child
+	// blocks, etc, in parallel.
+	respChans := make([]<-chan resp, 0, len(pblock.IPtrs))
+	eg, groupCtx := errgroup.WithContext(ctx)
+	nextBlockOffsetThisLevel := int64(-1)
+	for i, iptr := range pblock.IPtrs {
+		// Some byte of this block is included in the left side of the
+		// range if `startOff` is less than the largest byte offset in
+		// the block.
+		inRangeLeft := true
+		if i < len(pblock.IPtrs)-1 {
+			inRangeLeft = startOff < pblock.IPtrs[i+1].Off
+		}
+		if !inRangeLeft {
+			continue
+		}
+		// Some byte of this block is included in the right side of
+		// the range if `endOff` is bigger than the smallest byte
+		// offset in the block (or if we're explicitly reading all the
+		// data to the end).
+		inRangeRight := endOff == -1 || endOff > iptr.Off
+		if !inRangeRight {
+			// This block is the first one past the offset range
+			// amount the children.
+			nextBlockOffsetThisLevel = iptr.Off
+			break
+		}
+
+		ptr := iptr.BlockPointer
+		childIndex := i
+		respCh := make(chan resp, 1)
+		respChans = append(respChans, respCh)
+		eg.Go(func() error {
+			block, _, err := fd.getter(
+				groupCtx, fd.kmd, ptr, fd.file, blockReadParallel)
+			if err != nil {
+				return err
+			}
+			// Recurse down to the level of the child.
+			pfr, blocks, nextBlockOffset, err := fd.getLeafBlocksForOffsetRange(
+				groupCtx, ptr, block, startOff, endOff, prefixOk)
+			if err != nil {
+				return err
+			}
+
+			// Append self to the front of every path.
+			var r resp
+			for _, p := range pfr {
+				newPath := append([]parentBlockAndChildIndex{{
+					pblock:     pblock,
+					childIndex: childIndex,
+				}}, p...)
+				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+			}
+			r.blocks = blocks
+			r.nextBlockOffset = nextBlockOffset
+			respCh <- r
+			return nil
+		})
+	}
+
+	err = eg.Wait()
+	// If we are ok with just getting the prefix, don't treat a
+	// deadline exceeded error as fatal.
+	if prefixOk && err == context.DeadlineExceeded {
+		err = nil
+	}
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	blocks = make(map[BlockPointer]*FileBlock)
+	minNextBlockOffsetChild := int64(-1)
+outer:
+	for _, respCh := range respChans {
+		select {
+		case r := <-respCh:
+			pathsFromRoot = append(pathsFromRoot, r.pathsFromRoot...)
+			for ptr, block := range r.blocks {
+				blocks[ptr] = block
+			}
+			if r.nextBlockOffset != -1 &&
+				r.nextBlockOffset < minNextBlockOffsetChild {
+				minNextBlockOffsetChild = r.nextBlockOffset
```

Comment:
We want to find the leftmost block that's to the right of the range, the one immediately following it.